tools/cell_census_builder/__main__.py

```diff
@@ -210,80 +197,123 @@ def build_step1_get_source_assets(args: argparse.Namespace, assets_path: str) ->
+        for e in experiment_builders:
+            e.create(census_data=root_collection[CENSUS_DATA_NAME])
+            logging.info(f"Experiment {e.name} will contain {e.n_obs} cells from {e.n_datasets} datasets")
+
+        logging.info("Build step 2 - axis creation - finished")
+        return root_collection
 
-    # Write obs axis and accumulate var axis (and remember the datasets that pass our filter)
+
+def filter_datasets(
+    assets_path: str, datasets: List[Dataset], experiment_builders: List[ExperimentBuilder]
+) -> List[Dataset]:
+    # Accumulate obs and var axes (and remember the datasets that pass our filter)
     filtered_datasets = []
     N = len(datasets) * len(experiment_builders)
     n = 1
+
     for dataset, ad in open_anndata(assets_path, datasets, backed="r"):
         dataset_total_cell_count = 0
         for e in experiment_builders:
-            dataset_total_cell_count += e.accumulate_axes(dataset, ad, progress=(n, N))
+            logging.info(f"{e.name}: accumulate axis for dataset '{dataset.dataset_id}' ({n} of {N})")
+            if ad_filtered := e.filter_anndata_cells(ad):
+                dataset_total_cell_count += ad_filtered.shape[0]
             n += 1
 
         dataset.dataset_total_cell_count = dataset_total_cell_count
         if dataset_total_cell_count > 0:
             filtered_datasets.append(dataset)
+    return filtered_datasets
 
-    # Commit / write var
-    for e in experiment_builders:
-        e.commit_axis()
-        logging.info(f"Experiment {e.name} will contain {e.n_obs} cells from {e.n_datasets} datasets")
 
-    logging.info("Build step 2 - axis creation - finished")
-    return top_level_collection, filtered_datasets
+def accumulate_axes(assets_path: str, datasets: List[Dataset], experiment_builders: List[ExperimentBuilder]) -> int:
+    N = len(datasets) * len(experiment_builders)
+    n = 1
+    dataset_total_cell_count = 0
 
+    # append to `obs`; accumulate `var` data
+    for dataset, ad in open_anndata(assets_path, datasets, backed="r"):
+        for eb in experiment_builders:
+            logging.info(f"{eb.name}: accumulate axis for dataset '{dataset.dataset_id}' ({n} of {N})")
+            dataset_total_cell_count += eb.accumulate_axes(dataset, ad)
+            n += 1
+
+    # populate `var`; create empty `presence` now that we have its dimensions
+    if len(datasets) > 0:
+        for eb in experiment_builders:
+            eb.populate_var_axis()
+
+            # SOMA does not currently support empty arrays, so special case this corner-case.
+            if eb.n_var > 0:
+                max_dataset_joinid = max(d.soma_joinid for d in datasets)
+                eb.experiment.ms["RNA"].add_new_sparse_ndarray(
+                    FEATURE_DATASET_PRESENCE_MATRIX_NAME, type=pa.bool_(), shape=(max_dataset_joinid + 1, eb.n_var)
+                )
+
+    return dataset_total_cell_count
 
-def build_step3_create_X_layers(
+
+def build_step3_populate_axes_and_X_layers(
```

Comment:
this function opens each AnnData twice, which isn't really necessary.  with the soma_joinids already defined, you could refactor this so that:
* for each AnnData, have a sub-process:
    1. open the full AnnData (backed=None)
    2. append obs, var and X[raw]
There is no race as there are no duplicate coordinates.

Reply:
In addition, you could process all AnnData concurrently with this model, or better yet, do all (experiment, AnnData) pairs concurrently.

With your change (which I like), only `build_step2` is single-threaded.

Reply:
I refactored to fold together the iteration through the datasets (anndatas), so that the dataset/cell filtering step and axes processing is done with a single, serialized pass.

I'm not sure we can parallelize this though due to the var accumulation.  Even if it can be, let's save that for another PR!

Reply:
OK, new changes to the build phases look good to me. And you are right about `var` accumulation - I "spaced" on that not-so-minor detail....